File: movie_generator/gpt_client.py
"""
ChatGPT client — generates Blender Python scripts from natural-language movie descriptions.
"""

import re
from typing import Optional, Tuple

import openai

# ---------------------------------------------------------------------------
# System prompt sent to GPT on every request
# ---------------------------------------------------------------------------
_SYSTEM_PROMPT = """\
You are an expert Blender 3D animation Python developer.
When the user describes a movie they want, you must:

1. Reply with ONE short sentence describing what you will create (no code here).
2. Then output a COMPLETE, self-contained Blender Python script inside a
   ```python ... ``` code block.

━━━━━━━━━━━  MANDATORY SCRIPT REQUIREMENTS  ━━━━━━━━━━━
• import bpy  (always first line)
• Clear the default scene:
      bpy.ops.wm.read_factory_settings(use_empty=True)
• Set render engine to BLENDER_EEVEE  (fast real-time renderer)
• Set output path EXACTLY to: {output_path}
• Set file format:
      scene.render.image_settings.file_format = 'FFMPEG'
      scene.render.ffmpeg.format = 'MPEG4'
      scene.render.ffmpeg.codec = 'H264'
      scene.render.ffmpeg.constant_rate_factor = 'HIGH'
• Frame rate: 24 fps
• Resolution: 1280 × 720  (unless the user asks for something else)
• Set frame_start and frame_end to match the requested length
  (e.g. 10 s → frames 1–240)
• Create ALL scene content programmatically:
    – At least one camera (set as scene.camera)
    – At least one light (Sun, Point, or Area)
    – All 3-D objects, materials, textures, and keyframe animations
• Materials MUST use nodes  (mat.use_nodes = True)
• NO external file imports — everything built from scratch with bpy
• Very last line of the script:
      bpy.ops.render.render(animation=True)

Write ONLY valid Python / bpy code inside the code block.
Do NOT add explanations inside the code block.
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
"""


class GPTClient:
    """Wraps the OpenAI Chat Completions API for movie-script generation."""

    def __init__(self, api_key: str, model: str = "gpt-4o"):
        self._client = openai.OpenAI(api_key=api_key)
        self._model = model
        # Keep full conversation so GPT remembers earlier context
        self._history: list[dict] = []

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def generate_blender_script(
        self, user_request: str, output_path: str
    ) -> Tuple[Optional[str], str]:
        """
        Ask GPT to create a Blender script for *user_request*.
        Returns (script_code, description_text).
        """
        system = _SYSTEM_PROMPT.replace("{output_path}", output_path)

        user_msg = (
            f"Create a movie for me:\n\n{user_request}\n\n"
            f"Output file path: {output_path}"
        )
        self._history.append({"role": "user", "content": user_msg})

        reply = self._chat(system)

        script = self._extract_code(reply)
        description = self._extract_description(reply)
        return script, description

    def fix_script(
        self, broken_script: str, error_text: str, output_path: str
    ) -> Tuple[Optional[str], str]:
        """
        Ask GPT to fix a broken Blender script based on the error message.
        Returns (fixed_script_code, reply_text).
        """
        system = _SYSTEM_PROMPT.replace("{output_path}", output_path)

        fix_msg = (
            "The Blender script failed with this error:\n\n"
            f"```\n{error_text[:3000]}\n```\n\n"
            "Here is the original script:\n\n"
            f"```python\n{broken_script}\n```\n\n"
            f"Output path (keep it unchanged): {output_path}\n\n"
            "Please provide a completely corrected script."
        )
        self._history.append({"role": "user", "content": fix_msg})

        reply = self._chat(system)
        script = self._extract_code(reply)
        return script, reply

    def reset(self):
        """Clear conversation history (start fresh)."""
        self._history.clear()

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _chat(self, system_prompt: str) -> str:
        response = self._client.chat.completions.create(
            model=self._model,
            messages=[
                {"role": "system", "content": system_prompt},
                *self._history,
            ],
            temperature=0.7,
            max_tokens=4096,
        )
        reply = response.choices[0].message.content or ""
        self._history.append({"role": "assistant", "content": reply})
        return reply

    @staticmethod
    def _extract_code(text: str) -> Optional[str]:
        # Prefer ```python ... ``` blocks
        for pattern in (r"```python\s*(.*?)\s*```", r"```\s*(.*?)\s*```"):
            m = re.search(pattern, text, re.DOTALL)
            if m:
                code = m.group(1).strip()
                if "import bpy" in code:
                    return code
        return None

    @staticmethod
    def _extract_description(text: str) -> str:
        # Strip code blocks and return the first non-empty line
        clean = re.sub(r"```.*?```", "", text, flags=re.DOTALL).strip()
        lines = [ln.strip() for ln in clean.splitlines() if ln.strip()]
        return lines[0] if lines else "Generating your movie..."

File: android_movie_generator/app/src/main/java/com/moviegenerator/GptClient.kt
package com.moviegenerator

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

private val SYSTEM_PROMPT = """
You are an expert Three.js JavaScript 3D animation developer for a mobile movie generator app.

When the user describes a movie they want, respond with:
1. ONE short sentence describing what you will create.
2. A complete JavaScript snippet inside a ```javascript ... ``` code block.

━━━━━━━  MANDATORY CODE REQUIREMENTS  ━━━━━━━
The script runs inside a Three.js r158 WebGL context where `THREE`, `scene`,
`camera`, and a WebGLRenderer are already set up at 720×1280 portrait.

Your code MUST define exactly these three things:

  const TOTAL_FRAMES = <integer>;          // frames at 24 fps, e.g. 10 s = 240

  function initScene(THREE, scene, camera) {
    // Create objects, lights, materials.  Store them in module-scope variables.
    scene.background = new THREE.Color(0x000010);  // example — always set bg
  }

  function updateScene(frameNum, THREE, scene, camera) {
    // Animate for frame `frameNum` (0 … TOTAL_FRAMES-1).
    // Update positions, rotations, colors, etc.
  }

Rules:
• Use ONLY Three.js r158 API — THREE is already in scope, no imports needed.
• Declare shared objects (meshes, lights, etc.) with let/const at the top of
  the code block so both initScene and updateScene can access them.
• Always add at least AmbientLight + DirectionalLight.
• Use MeshStandardMaterial or MeshPhongMaterial for visible objects.
• Do NOT use any loader, external texture URL, or require().
• Keep code self-contained and runnable — no syntax errors.
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
""".trimIndent()

class GptClient(private val apiKey: String) {

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .build()

    private val history = mutableListOf<JSONObject>()

    /** Returns Pair(animationCode, shortDescription) or throws on error. */
    fun generateAnimation(userRequest: String): Pair<String, String> {
        history.add(userMessage("Create a movie for me: $userRequest"))
        val reply = chat()
        val code = extractCode(reply) ?: error("No JavaScript code block found in GPT reply.")
        val desc = extractDescription(reply)
        return Pair(code, desc)
    }

    fun fixAnimation(brokenCode: String, errorMsg: String): Pair<String, String> {
        val fix = "The Three.js code failed:\n\n```\n${errorMsg.take(1500)}\n```\n\n" +
                "Original code:\n```javascript\n$brokenCode\n```\n\nPlease fix it."
        history.add(userMessage(fix))
        val reply = chat()
        val code = extractCode(reply) ?: error("No code block in fix reply.")
        val desc = extractDescription(reply)
        return Pair(code, desc)
    }

    fun reset() = history.clear()

    // ── private ───────────────────────────────────────────────────────────────

    private fun chat(): String {
        val messages = JSONArray().apply {
            put(JSONObject().put("role", "system").put("content", SYSTEM_PROMPT))
            history.forEach { put(it) }
        }
        val body = JSONObject()
            .put("model", "gpt-4o")
            .put("messages", messages)
            .put("temperature", 0.7)
            .put("max_tokens", 4096)
            .toString()
            .toRequestBody("application/json".toMediaType())

        val req = Request.Builder()
            .url("https://api.openai.com/v1/chat/completions")
            .addHeader("Authorization", "Bearer $apiKey")
            .post(body)
            .build()

        val response = client.newCall(req).execute()
        val responseBody = response.body?.string() ?: error("Empty response from OpenAI")
        if (!response.isSuccessful) error("OpenAI error ${response.code}: $responseBody")

        val content = JSONObject(responseBody)
            .getJSONArray("choices")
            .getJSONObject(0)
            .getJSONObject("message")
            .getString("content")

        history.add(JSONObject().put("role", "assistant").put("content", content))
        return content
    }

    private fun userMessage(text: String) =
        JSONObject().put("role", "user").put("content", text)

    private fun extractCode(text: String): String? {
        val patterns = listOf(
            Regex("```javascript\\s*(.*?)\\s*```", RegexOption.DOT_MATCHES_ALL),
            Regex("```js\\s*(.*?)\\s*```", RegexOption.DOT_MATCHES_ALL),
            Regex("```\\s*(.*?)\\s*```", RegexOption.DOT_MATCHES_ALL),
        )
        for (p in patterns) {
            val m = p.find(text)
            if (m != null) {
                val code = m.groupValues[1].trim()
                if ("TOTAL_FRAMES" in code) return code
            }
        }
        return null
    }

    private fun extractDescription(text: String): String {
        val clean = text.replace(Regex("```.*?```", RegexOption.DOT_MATCHES_ALL), "").trim()
        return clean.lines().firstOrNull { it.isNotBlank() }?.trim() ?: "Generating your movie…"
    }
}
